File: backend/app/schemas/nick_live.py
from datetime import datetime

from pydantic import BaseModel, Field


class UserPayload(BaseModel):
    id: int
    name: str = Field(max_length=100)
    shop_id: int | None = None
    avatar: str | None = Field(default=None, max_length=500)


class NickLiveCreate(BaseModel):
    """User pastes JSON with user object + cookies string"""

    user: UserPayload
    cookies: str = Field(min_length=1, max_length=20000)


class NickLiveResponse(BaseModel):
    id: int
    name: str
    user_id: int
    shop_id: int | None
    avatar: str | None
    created_at: datetime
    model_config = {"from_attributes": True}


class LiveSession(BaseModel):
    sessionId: int
    title: str
    coverImage: str
    startTime: int
    duration: int
    status: int
    views: int
    viewers: int
    peakViewers: int
    comments: int
    # status=1 and duration=0 means currently live


class LiveSessionsResponse(BaseModel):
    sessions: list[LiveSession]
    active_session: LiveSession | None  # the one with duration=0 and status=1


class CommentItem(BaseModel):
    id: str | None = None
    userName: str | None = None
    username: str | None = None
    nick_name: str | None = None
    nickname: str | None = None
    content: str | None = None
    comment: str | None = None
    message: str | None = None
    msg: str | None = None
    timestamp: int | None = None
    create_time: int | None = None
    ctime: int | None = None


class ScanStatus(BaseModel):
    is_scanning: bool
    session_id: int | None = None
    comment_count: int = 0


# --- Moderator schemas ---


class ModeratorSaveCurlRequest(BaseModel):
    """Save moderator cURL template for a nick_live."""

    curl_text: str = Field(min_length=10, max_length=50000)


class ModeratorReplyRequest(BaseModel):
    """Send a reply to a specific guest."""

    guest_name: str = Field(min_length=1, max_length=200)
    guest_id: int
    reply_text: str = Field(min_length=1, max_length=2000)


class ModeratorAutoReplyRequest(BaseModel):
    """Auto-reply to multiple comments."""

    comments: list[dict]
    reply_text: str = Field(min_length=1, max_length=2000)


class ModeratorStatus(BaseModel):
    """Whether moderator is configured for a nick_live."""

    nick_live_id: int
    configured: bool
    host_id: str | None = None
    has_usersig: bool = False


# --- Host schemas ---


class HostGetCredentialsResponse(BaseModel):
    status: str
    uuid: str | None = None
    error: str | None = None


class HostConfigStatus(BaseModel):
    configured: bool
    uuid: str | None = None
    has_usersig: bool = False
    proxy: str | None = None


class AutoPostStartRequest(BaseModel):
    session_id: int


class AutoPostStatusResponse(BaseModel):
    running: bool


class HostPostRequest(BaseModel):
    """Manual host comment (type 101)."""

    content: str = Field(..., min_length=1, max_length=2000)
    session_id: int

File: frontend/src/api/nickLive.ts
import apiClient from "./client";

export interface NickLive {
  id: number;
  name: string;
  user_id: number;
  shop_id: number | null;
  avatar: string | null;
  created_at: string;
}

export interface LiveSession {
  sessionId: number;
  title: string;
  coverImage: string;
  startTime: number;
  duration: number;
  status: number;
  views: number;
  viewers: number;
  peakViewers: number;
  comments: number;
}

export interface LiveSessionsResponse {
  sessions: LiveSession[];
  active_session: LiveSession | null;
}

export interface ScanStatus {
  is_scanning: boolean;
  session_id: number | null;
  comment_count: number;
}

export interface CommentItem {
  id?: string;
  userName?: string;
  username?: string;
  nick_name?: string;
  nickname?: string;
  content?: string;
  comment?: string;
  message?: string;
  msg?: string;
  timestamp?: number;
  create_time?: number;
  ctime?: number;
  streamerId?: number;
  userId?: number;
  user_id?: number;
  uid?: number;
}

export async function createNickLive(data: {
  user: Record<string, unknown>;
  cookies: string;
}): Promise<NickLive> {
  const res = await apiClient.post("/nick-lives", data);
  return res.data;
}

export async function listNickLives(): Promise<NickLive[]> {
  const res = await apiClient.get("/nick-lives");
  return res.data;
}

export async function deleteNickLive(id: number): Promise<void> {
  await apiClient.delete(`/nick-lives/${id}`);
}

export async function getSessions(
  id: number
): Promise<LiveSessionsResponse> {
  const res = await apiClient.get(`/nick-lives/${id}/sessions`);
  return res.data;
}

export async function startScan(
  id: number,
  sessionId: number
): Promise<void> {
  await apiClient.post(
    `/nick-lives/${id}/scan/start?session_id=${sessionId}`
  );
}

export async function stopScan(id: number): Promise<void> {
  await apiClient.post(`/nick-lives/${id}/scan/stop`);
}

export async function getScanStatus(id: number): Promise<ScanStatus> {
  const res = await apiClient.get(`/nick-lives/${id}/scan/status`);
  return res.data;
}

export async function getComments(id: number): Promise<CommentItem[]> {
  const res = await apiClient.get(`/nick-lives/${id}/comments`);
  return res.data;
}

// --- Moderator API ---

export interface ModeratorStatus {
  nick_live_id: number;
  configured: boolean;
  host_id: string | null;
  has_usersig: boolean;
}

export interface ModeratorReplyResult {
  success: boolean;
  status_code?: number;
  response?: string;
  guest?: string;
  reply?: string;
  error?: string;
}

export async function saveModeratorCurl(
  nickLiveId: number,
  curlText: string
): Promise<{ nick_live_id: number; host_id: string; status: string }> {
  const res = await apiClient.post(
    `/nick-lives/${nickLiveId}/moderator/save-curl`,
    { curl_text: curlText }
  );
  return res.data;
}

export async function getModeratorStatus(
  nickLiveId: number
): Promise<ModeratorStatus> {
  const res = await apiClient.get(
    `/nick-lives/${nickLiveId}/moderator/status`
  );
  return res.data;
}

export async function removeModerator(nickLiveId: number): Promise<void> {
  await apiClient.delete(`/nick-lives/${nickLiveId}/moderator`);
}

export async function sendModeratorReply(
  nickLiveId: number,
  guestName: string,
  guestId: number,
  replyText: string
): Promise<ModeratorReplyResult> {
  const payload = {
    guest_name: guestName,
    guest_id: guestId,
    reply_text: replyText,
  };
  console.log("[sendModeratorReply] payload →", payload);
  const res = await apiClient.post(
    `/nick-lives/${nickLiveId}/moderator/reply`,
    payload
  );
  console.log("[sendModeratorReply] response →", res.data);
  return res.data;
}

export async function autoReplyComments(
  nickLiveId: number,
  comments: CommentItem[],
  replyText: string
): Promise<ModeratorReplyResult[]> {
  const res = await apiClient.post(
    `/nick-lives/${nickLiveId}/moderator/auto-reply`,
    {
      comments,
      reply_text: replyText,
    }
  );
  return res.data;
}
